164 allow ip based policy enforcement

jrajahalme opened the pull request:
These commits implement the "ToCIDR" and "FromCIDR" policies. Legacy kernels use inline arrays and a linear search, kernels with working LPM trie maps use them instead. As future work we need to add a limit on the number of CIDRs we try to inline in bfs, and enhance "cilium policy trace" to also trace CIDR policies.

jrajahalme committed:
bpf: More rigorous bpf map probes.

Detect kernel bug wherein bpf map flags were not properly stored.  Do
this by importing bpf map selftest functionality from iproute2.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
policy: Add L3 rule validation and unit tests.

Verify that strings given to ToCIDR and FromCIDR successfully parse as
CIDRs and have a non-zero prefix length.  Plain IPs are interpreted as
fully masked addresses, or in case of IPv4 with all zeroes after a
subnetwork address, as subnetwork addresses.

Add basic unit tests to check CIDR rule validation.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
pkg/maps: Add support for flags in map creation.

Longest Prefix Match (LPM) trie map creation requires non-zero flags.
Add support for flags in the BPF map creation.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
pkg/policy: Add L3 support.

Generate both LPM trie maps (if kernel supports them) and inline
tables for the L3 prefixes so that datapath can enforce L3 CIDR
policies.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

aanm commented on pkg/policy/rule_test.go:
```diff
@@ -149,3 +149,60 @@ func (ds *PolicyTestSuite) TestL4Policy(c *C) {
+	// Must be parsable
+	err = api.Rule{
+		Ingress: []api.IngressRule{{
+			FromCIDR: []api.CIDR{{IP: "10.0.1..0/24"}},
```
Are the 2 dots in the IP intencional?

tgraf replied:
It shouldn't pass the `Validate()` so if it does, that must be a bug.

aanm commented on pkg/endpoint/policy.go:
```diff
@@ -240,6 +240,32 @@ func (e *Endpoint) regenerateConsumable(owner Owner) (bool, error) {
 
+// Must be called with endpointsMU held
+func (e *Endpoint) regenerateL3Policy(owner Owner) (bool, error) {
+	c := e.Consumable
```
can't `e.Consumable` be `nil` here?

tgraf replied:
We only ever call `regeneratePolicy()` when `e.Consumable != nil`. Let's add a comment to `regeneratePolicy()` to make this obvious.

jrajahalme replied:
No, since the caller in regenerate() calls e.regeneratePolicy() only if e.Consumable is not nil.

tgraf commented on bpf/bpf_lxc.c:
```diff
@@ -305,6 +305,7 @@ static inline int ipv6_l3_from_lxc(struct __sk_buff *skb,
 		return ipv6_local_delivery(skb, l3_off, l4_off, SECLABEL, ip6, tuple->nexthdr);
 	} else {
 #ifdef LXC_NAT46
+	  	/* XXX: Is all NATted egress traffic allowed? */
```
If a container emits IPv6 which then gets NAT64 to IPv4, I would expect that egress CIDR rules get applied to the translated IPv4 packets.

tgraf replied:
I'm fine with excluding this use case initially though.

jrajahalme replied:
Looking into this deeper, it seem this case is already covered, as follows:
1. NATted egress traffic tail call into tail_ipv6_to_ipv4()
2. tail_ipv6_to_ipv4() tail calls into tail_handle_ipv4(), which calls handle_ipv4()
3. handle_ipv4() already checks for the egress CIDR policies if outgoing traffic is not otherwise allowed.

Based on this I removed the comment.

aanm commented on pkg/endpoint/policy.go:
```diff
@@ -240,6 +240,32 @@ func (e *Endpoint) regenerateConsumable(owner Owner) (bool, error) {
+
+	c.Mutex.RLock()
+	ctx := policy.SearchContext{
+		To:    c.LabelList,
```
Needs to be `c.LabelList.Labels.DeepCopy()` or RUnlock after line 260.

tgraf replied:
I'm having some issues with all the `DeepCopy()`. While it avoids changes under a live pointer, it does not solve the race itself. After the Unlock, the LabelList can change and we calculate policy for stale labels without knowing it.

I agree with the `DeepCopy()` fix short term but we need to look at this very closely and make sure we are providing accurate guarantees regarding the policy that is generated.

[*Jarno*]: For some reason I can't reply to this but can edit instead?

Rather than using DeepCopy() I preferred to keep the locks longer. Please note that regenerateConsumable() has the same issue and I did not touch it for now.

jrajahalme replied:
Does regenerateConsumable() (right above) have the same issue then?

jrajahalme replied:
Also, what is the locking order between c.Mutex and repo.Mutex?

jrajahalme replied:
I see that e.Mutex is already held during regeneration.

tgraf replied:
repo.Mutex must always be taken before c.Mutex. We should add a comment.

aanm commented on pkg/policy/l3.go:
```diff
@@ -0,0 +1,170 @@
+// L3PolicyMap is a list of CIDR filters indexable by address/prefixlen
+// key format: "address/prefixlen", e.g., "10.1.1.0/24"
+type L3PolicyMap struct {
+	Map         map[string]net.IPNet // Allowed L3 prefixes
```
Is this map accessed by multiple threads? If yes needs a lock on all R/W operations

tgraf replied:
We have an `L3PolicyMap` per endpoint which would always take the `e.Mutex`.

tgraf replied:
I think it makes sense to leave a note in the comment of `L3PolicyMap` to indicate that locking is in the responsibility of the whoever uses the struct.

jrajahalme replied:
Regeneration currently always creates a new L3Policy and actually does not compare the new one to old one. BPF regeneration uses the endpoint.L3Policy, but that is also run from the same thread. endpoint.DeepCopy() could potentially be called from another thread in parallel, but it is only used by the unit tests.

In summary, it seems to me that each L3Policy is accessed from one thread only and further locking is not needed.

tgraf commented on bpf/lib/maps.h:
```diff
@@ -77,6 +78,180 @@ struct bpf_elf_map __section_maps CALLS_MAP = {
+	.size_key	= sizeof(struct bpf_lpm_trie_key4),
+	.size_value	= sizeof(struct policy_entry),
+	.pinning	= PIN_GLOBAL_NS,
+	.max_elem	= 1024,
```
Can we make `.max_elem` based on a macro constant? We could do something like:
```
#ifndef CIDR4_INGRESS_MAP_SIZE
#define CIDR4_INGRESS_MAP_SIZE 1024
#endif
```

Which would allow to use change the size of the map at any time from the user space side without ever touching the datapath again.

jrajahalme replied:
I adopted your suggestion but changed the macro name to LPM_MAP_SIZE and reused it for all four maps (v4/v6, ingress/egress).

tgraf commented on bpf/lib/maps.h:
```diff
@@ -77,6 +78,180 @@ struct bpf_elf_map __section_maps CALLS_MAP = {
+	const int size = (sizeof(map) / sizeof(map[0]));
+	int i;
+
+#pragma unroll
```
Does the `#pragma unroll` work here or will this require the same fixes as @jrfastab is working on in #882?

jrajahalme replied:
Seem to ge working for now, but will have to see when this code is exposed to more testing in different kernels, llvm versions, etc.

tgraf commented on pkg/maps/cidrmap/cidrmap.go:
```diff
@@ -0,0 +1,131 @@
+	ones, _ := cidr.Mask.Size()
+	key.Prefixlen = uint32(ones)
+	// IPv4 address can be represented by 16 byte slice in 'cidr.IP',
+	// in which case the address is at the end of the slice.
```
Thanks for the comment here. This would not have been obvious to me otherwise.

tgraf commented on pkg/maps/cidrmap/cidrmap.go:
```diff
@@ -0,0 +1,131 @@
+
+func (cm *CIDRMap) AllowCIDR(cidr net.IPNet) error {
+	key := cm.CIDRKeyInit(cidr)
+	entry := PolicyEntry{Action: 1}
```
Can we add a comment there to indicate that the `Action` does not matter for now?

jrajahalme replied:
Currently even the policy entry counters are not being updated, as LPM maps can not even be dumped. I'll change the code to not use the policy entry at all to make this clearer.

tgraf commented on pkg/endpoint/endpoint.go:
```diff
@@ -277,6 +280,7 @@ func (e *Endpoint) GetModel() *models.Endpoint {
 			IPV4: e.IPv4.String(),
 			IPV6: e.IPv6.String(),
 		},
+		//		L3Policy:         e.L3Policy.GetModel(),
```
This seems to be a leftover

jrajahalme replied:
Yep, removed, thanks!

tgraf commented on pkg/endpoint/l3maps.go:
```diff
@@ -0,0 +1,69 @@
+	os.RemoveAll(path)
+}
+
+func (l3 *L3Maps) CreateBpfMap(mt L3MapType, path string) error {
```
Should we rename this to `ResetBpfMap`? I was initially confused that this would be a one-time operation.

jrajahalme replied:
Good suggestion, thanks!

jrajahalme committed:
bpf: Support L3/IP access policy.

L3 IP policy can allow traffic if the policy would otherwise drop them.

Kernels with working LPM trie maps will use them, otherwise the CIDR
prefixes are inlined into the bpf code.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
tests: Show which test is running.

Test failures are easier to diagnose if we know which test failed.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
bpf: Add probe for LPM trie map type.

Add probe for longest prefix match trie map type added in Linux 4.11.
LPM map requires the flag BPF_F_NO_PREALLOC, so we add flags to struct
bpf_map_fixups, which defaults to zero for all other map types.

Also add debug output in case map creation fails.

Tested with Linux 4.9 (probe fails) and 4.12-rc1 (probe succeeds).

Signed-off-by Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
pkg/maps: Add new cidrmap.

CIDR maps are backed by in-kernel longest prefix match trie maps. LPM
maps do not yet support element deletion or dumping, so those are not
implemented in the go interface either.

The entry value is also currenlty unused, as we allow traffic based on
the existence of the entry, not by it's value.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>
